proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,590 @@
+  return loadById(id)
+}
+
+// Regular Kotlin
+fun <A> fetchById(id: Int, repository: Repository<A>): A? =
+  with (repository) {
+    return loadById(id)
+  }
+```
+
+On the call site, we could use it as follows:
+
+```kotlin
+fetchById<User>(11829) // compiles since we got evidence of a `Repository<User>` in scope.
+fetchById<Coin>(12398) // does not compile: No `Repository<Coin>` evidence defined in scope!
+```
+
+All values can be passed to functions with extension parameters, or we can omit extension parameters and let the compiler resolve the suitable extensions for them. This makes the approach really similar to the way default arguments work in Kotlin.
+
+```kotlin
+fetchById<User>(11829) // compiles since we got evidence of a `Repository<User>` in scope.
+fetchById<User>(11829, UserRepository()) // you can provide it manually.
+fetchById<User>(11829, repository=UserRepository()) // you can provide it manually with named application.
+```
+
+When `with` is used in class constructors, it is important to **add val to extension class fields** to make sure they are accessible in the scope of the class. Here, the `with` keyword adds the value to the scope of every method in the class. To showcase this, let's say we have a `Validator<A>`, like:
+
+```kotlin
+interface Validator<A> {
+  fun A.isValid(): Boolean
+}
+```
+
+In this scenario, the following classes would be equivalent:
+
+```kotlin
+// Kotlin + KEEP-87
+extension class ValidatedRepository<A>(with val V: Validator<A>) : Repository<A> {
+
+    val storedUsers: MutableMap<Int, A> = mutableMapOf() // e.g. users stored in a DB
+
+    override fun loadAll(): Map<Int, A> {
+        return storedUsers.filter { it.value.isValid() }
+    }
+
+    override fun loadById(id: Int): A? {
+        return storedUsers[id]?.let { if (it.isValid()) it else null }
+    }
+
+    override fun A.save() {
+        storedUsers[generateKey(this)] = this
+    }
+}
+
+// Regular Kotlin
+class ValidatedRepository<A>(val V: Validator<A>) : Repository<A> {
+
+    val storedUsers: MutableMap<Int, A> = mutableMapOf() // e.g. users stored in a DB
+
+    override fun loadAll(): Map<Int, A> {
+        with (V) {
+            return storedUsers.filter { it.value.isValid() }
+        }
+    }
+
+    override fun loadById(id: Int): A? {
+        with (V) {
+            return storedUsers[id]?.let { if (it.isValid()) it else null }
+        }
+    }
+
+    override fun A.save() {
+        storedUsers[generateKey(this)] = this
+    }
+}
+```
+
+As you can see in the first example, `A.isValid()` becomes available inside the method’s scope automatically. The equivalent version of doing this without KEEP-87 is to manually add `with (V)` inside each method, as you can see in the second example.  
+
+## Composition and chain of evidences
+
+Constraint interface declarations and extension evidences can encode further constraints on their type parameters so that they can be composed nicely:
```

Comment:
Unless there are other mechanisms in place for ensuring termination of the resolver, I think extension classes with constraints need to be constrained. For example, if I have
```kotlin
extension interface Foo<A> {}
extension class Bar(with val fooA: Foo<A>): Foo<A> {}
```
then a naive resolver will run forever. In Haskell, I think the restriction is that the types mentioned in each constraint must be a strict subterm of one of the types mentioned in the goal, or the constraint must refer to a type classe with no cyclic dependency on the goal.